File: recipiece-ui/src/app/api/model/recipe.ts
import {Model} from './model';
import {IEquipment} from './equipment';

export interface IRecipe extends Model {
  _id?: string;
  name: string;
  description: string;
  private: boolean;
  steps: IRecipeStep[];
  ingredients: IRecipeIngredient[];
  equipment?: IEquipment[],
  tags: string[];
  links?: string[];
  owner?: string;
  advanced: IRecipeAdvancedOptions;
}

type StepLenUnit = 'm' | 'h' | 'd' | 'w' | 's';


export interface IRecipeAdvancedOptions {
  highAltitude?: boolean;
  altitude?: number;
}


export interface IRecipeStep {
  content: string;
  length: {
    time: number;
    unit: StepLenUnit;
  };
}

export interface IRecipeIngredient {
  name: string;
  amount: string;
  unit?: string;
}

File: recipiece-server/Database/Models/Recipe.py
from typing import Optional, List, TypedDict

from Database.Models import BaseModel


class RecipeStepLength(TypedDict):
    time: int
    unit: str


class RecipeStep(TypedDict):
    content: str
    length: RecipeStepLength


class RecipeIngredient(TypedDict, total=False):
    name: str
    amount: str
    unit: Optional[str]


class RecipeAdvancedOptions(TypedDict, total=False):
    highAltitude: Optional[bool]
    altitude: Optional[int]


class Recipe(BaseModel.UserOwnedModel, total=False):
    name: str
    description: str
    private: bool
    steps: List[RecipeStep]
    ingredients: List[RecipeIngredient]
    links: Optional[List[str]]
    advanced: Optional[RecipeAdvancedOptions]
    tags: List[str]
